gluten-core/src/main/scala/io/glutenproject/extension/columnar/PullOutPreProject.scala

```diff
@@ -83,38 +79,53 @@ case class PullOutPreProject(session: SparkSession)
+  }
 
-        // The output of the sort operator is the same as the output of the child, therefore it
-        // is necessary to retain the output columns of the child in the pre-projection, and
-        // then add the expressions that need to be evaluated in the sortOrder. Finally, in the
-        // post-projection, the additional columns need to be removed, leaving only the original
-        // output of the child.
-        val preProject = ProjectExec(
-          eliminateProjectList(sort.child.outputSet, expressionMap.values.toSeq),
-          sort.child)
+  def applyForValidation[T <: SparkPlan](plan: T): T =
+    applyLocally
+      .lift(plan)
+      .map {
+        case p: ProjectExec => p.child
+        case other => other
+      }
+      .getOrElse(plan)
+      .asInstanceOf[T]
 
-        val newSort = sort.copy(sortOrder = newSortOrder, child = preProject)
-        newSort.copyTagsFrom(sort)
+  val applyLocally: PartialFunction[SparkPlan, SparkPlan] = {
+    case sort: SortExec if needsPreProject(sort) =>
+      val expressionMap = new mutable.HashMap[Expression, NamedExpression]()
+      val newSortOrder = getNewSortOrder(sort.sortOrder, expressionMap)
 
-        // The pre-project and post-project of SortExec always appear together, so it's more
-        // convenient to handle them together. Therefore, SortExec's post-project will no longer
-        // be pulled out separately.
-        ProjectExec(sort.child.output, newSort)
+      // The output of the sort operator is the same as the output of the child, therefore it
+      // is necessary to retain the output columns of the child in the pre-projection, and
+      // then add the expressions that need to be evaluated in the sortOrder. Finally, in the
+      // post-projection, the additional columns need to be removed, leaving only the original
+      // output of the child.
+      val preProject = ProjectExec(
+        eliminateProjectList(sort.child.outputSet, expressionMap.values.toSeq),
+        sort.child)
 
-      case topK: TakeOrderedAndProjectExec if needsPreProject(topK) =>
-        val expressionMap = new mutable.HashMap[Expression, NamedExpression]()
-        val newSortOrder = getNewSortOrder(topK.sortOrder, expressionMap)
-        val preProject = ProjectExec(
-          eliminateProjectList(topK.child.outputSet, expressionMap.values.toSeq),
-          topK.child)
-        val newTopK = topK.copy(sortOrder = newSortOrder, child = preProject)
-        newTopK.copyTagsFrom(topK)
-        newTopK
-    }
+      val newSort = sort.copy(sortOrder = newSortOrder, child = preProject)
+      newSort.copyTagsFrom(sort)
+
+      // The pre-project and post-project of SortExec always appear together, so it's more
+      // convenient to handle them together. Therefore, SortExec's post-project will no longer
+      // be pulled out separately.
+      val postProject = ProjectExec(sort.child.output, newSort)
+      postProject.setTagValue(PULLOUT_PROJECT_TAG, true)
```

Comment:
Can we do `unsetTagValue` original plan manutally after pulling out project ? Better to avoid introducing new tag.

Reply:
Do you mean inserting `AddTransformHintRules` again in `preOverrides` after `PullOutPreProject`?

Reply:
After the plan transformation is completed, it's difficult to identify which projects are pulled out projects.

Reply:
why do we need to know whether a project is a pull out project ?

Reply:
@ulysses-you When pulling out a project, because there is no tag, Spark will copy the tag of the "sort" to the pulled-out project. We need to find this project in order to unset the tag values in it.

We can also unset the `TRANSFORM_SUPPORTED` tag for sort, but this will require an additional validation for sort. Is that what you mean?

Reply:
> Can we do unsetTagValue original plan manutally after pulling out project ? Better to avoid introducing new tag.

I mean we can unsetTagValue for sort after we pulling out project, so there is no tag Spark would copy.